src/service/xhr-impl.js

```diff
@@ -197,6 +218,40 @@ export class Xhr {
       assertSuccess(response);
     });
   }
+
+  /**
+   * Add "__amp_source_origin" query parameter to the URL. Ideally, we'd be
+   * able to set a header (e.g. AMP-Source-Origin), but this will force
+   * preflight request on all CORS request.
+   * @param {!Window} win
+   * @param {string} url
+   * @return {string}
+   */
+  getCorsUrl(win, url) {
```

Comment:
We need to get the actual, full URL that is being sent by XHR, post all modifications.  We need it b/c we're relying on the browser cache to have the document in the case where the returned doc isn't valid AMP, so we render in cross-domain iframe.  In that case, we use this URL as src and if we don't have the full URL, we'll miss the cache and incur another round-trip through the ad network server to re-fetch the creative.